### PR TITLE
Use `ollama` Python client within `OllamaLLM`

### PR DESCRIPTION
**Pull request by sdiazlor**
- Update requests with [ollama-python](https://github.com/ollama/ollama-python).
- Update with all valid parameters and values from https://github.com/ollama/ollama/blob/main/docs/modelfile.md
- Tests passed

### CODE REVIEW

File: src/distilabel/llm/ollama.py [hunk 1]
```diff
@@ -116,22 +163,23 @@ def model_name(self) -> str:
...
+        """Checks if the Ollama API is available."""
+        msg = f"Could not connect to Ollama at {self.OLLAMA_HOST}. Check https://github.com/ollama/ollama-python/tree/main for deployment guide."
         try:
-            response = request.urlopen(self.OLLAMA_HOST)
-            if response.getcode() != 200:
-                raise Exception
-        except Exception as e:
+            ollama.list()
+        except ollama.ResponseError as e:
             raise ValueError(msg) from e
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:171]: Here I would remove the `msg` variable as only used once, and directly add it within `raise ValueError(...)`

File: src/distilabel/llm/ollama.py [hunk 2]
```diff
@@ -116,22 +163,23 @@ def model_name(self) -> str:
...
-            self._text_generation_with_backoff(
-                prompt=[{"role": "user", "content": "hi"}], max_tokens=1
+            ollama.chat(
+                model=self.model,
+                messages=[{"role": "user", "content": "hi"}],
+                options={"num_predict": 1},
             )
-        except Exception as e:
+        except ollama.ResponseError as e:
             raise ValueError(msg) from e
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:183]: Same here with the `msg` variable

File: src/distilabel/llm/ollama.py [hunk 3]
```diff
@@ -116,22 +163,23 @@ def model_name(self) -> str:
...
     def _api_model_available(self):
+        """Checks if the Ollama model is available"""
         msg = f"Model {self.model} is not available. Run `ollama run {self.model}` to serve the model."
         try:
-            self._text_generation_with_backoff(
-                prompt=[{"role": "user", "content": "hi"}], max_tokens=1
+            ollama.chat(
+                model=self.model,
+                messages=[{"role": "user", "content": "hi"}],
+                options={"num_predict": 1},
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:180]: Isn't this a bit more time consuming than calling `ollama.list` and then checking whether the Ollama model is listed there?

File: src/distilabel/llm/ollama.py [hunk 4]
```diff
@@ -147,49 +195,37 @@ def _api_model_available(self):
...
-                    hdrs=response.getheaders(),
-                    fp=None,
-                )
+        """Generates text using the Ollama API with backoff."""
+        try:
+            response = ollama.chat(
+                model=self.model,
+                messages=prompt,
+                options={
+                    "num_predict": kwargs.get("max_new_tokens") or self.max_new_tokens,
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:204]:
```suggestion
                    "num_predict": self.max_new_tokens,
```

File: src/distilabel/llm/ollama.py [hunk 5]
```diff
@@ -147,49 +195,37 @@ def _api_model_available(self):
...
+                    "num_gqa": self.num_gqa,
+                    "num_gpu": self.num_gpu,
+                    "repeat_last_n": self.repeat_last_n,
+                    "repeat_penalty": self.repeat_penalty,
+                    "seed": self.seed,
+                    "stop": self.stop,
+                    "tfs_z": self.tfs_z,
+                },
+            )
+            return response
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:221]:
```suggestion
            return ollama.chat(
                model=self.model,
                messages=prompt,
                options={
                    "num_predict": kwargs.get("max_new_tokens") or self.max_new_tokens,
                    "temperature": self.temperature,
                    "top_p": self.top_p,
                    "top_k": self.top_k,
                    "mirostat": self.mirostat,
                    "mirostat_eta": self.mirostat_eta,
                    "mirostat_tau": self.mirostat_tau,
                    "num_ctx": self.num_ctx,
                    "num_gqa": self.num_gqa,
                    "num_gpu": self.num_gpu,
                    "repeat_last_n": self.repeat_last_n,
                    "repeat_penalty": self.repeat_penalty,
                    "seed": self.seed,
                    "stop": self.stop,
                    "tfs_z": self.tfs_z,
                },
            )
```

**Review by alvarobartt — COMMENTED** [review]
Thanks for the PR @sdiazlor! I've just included some minor comments, and also the unit tests are still missing, but everything LGTM, so feel free to ping me once addressed so that we can merge 👍🏻

**Comment by alvarobartt** [comment]
Also @sdiazlor could you please include a line in both `README.md` and `docs/index.md` with the explanation on what does the `ollama` extra contain? It should have been done as part of #250 but seems missing, here's a similar PR for reference #308, thanks in advance 🤗

**Comment by sdiazlor** [comment]
> Also @sdiazlor could you please include a line in both `README.md` and `docs/index.md` with the explanation on what does the `ollama` extra contain? It should have been done as part of #250 but seems missing, here's a similar PR for reference #308, thanks in advance 🤗

Yes, no problem :)

**Comment by sdiazlor** [comment]
@alvarobartt I've made some changes according to your feedback and updated the tests (let me know if you think that any other should be included)

File: src/distilabel/llm/ollama.py [hunk 6]
```diff
@@ -116,23 +163,22 @@ def model_name(self) -> str:
...
     def _api_model_available(self):
-        msg = f"Model {self.model} is not available. Run `ollama run {self.model}` to serve the model."
+        """Checks if the Ollama model is available"""
         try:
-            self._text_generation_with_backoff(
-                prompt=[{"role": "user", "content": "hi"}], max_tokens=1
-            )
-        except Exception as e:
-            raise ValueError(msg) from e
+            ollama.show(f"{self.model}")
```
Comment (alvarobartt) [inline src/distilabel/llm/ollama.py:177]:
```suggestion
            ollama.show(self.model)
```

**Comment by alvarobartt** [comment]
Hi @sdiazlor it seems that there's a conflict with `httpx` dependency coming from `argilla`, I think this will be / has been patched as of 1.23.0, could you double check and upgrade `argilla` to `>=1.23.0` if this is the case? Thanks in advance 🤗

**Comment by sdiazlor** [comment]
@alvarobartt Yes, I was also aware of that issue, although as argilla 1.23 has not been released, tests are still failing by the moment

**Pull request merged**